python/ray/exceptions.py

```diff
@@ -367,6 +367,32 @@ def __str__(self):
         )
 
 
+@PublicAPI
+class OutOfMemoryError(RayError):
+    """Indicates that the node is running out of memory and is close to full.
+
+    This is raised if the node is low on memory and tasks or actors are being
+    evicted to free up memory.
+    """
+
+    def __init__(self, message):
+        self.message = message
+
+    def __str__(self):
+        return self.message
+
+
+@PublicAPI
+class NodeDiedError(RayError):
+    """Indicates that the node is either dead or unreachable."""
+
+    def __init__(self, message):
+        self.message = message
```

Comment:
Hmm wouldn't it be better to set the message string here and pass any related metadata, similar to what we do for other errors?

We could generate this error from multiple sources in the backend, so it will be harder to keep the message consistent if we do that. It's also a bit easier to read if we keep the message here.

Reply:
I looked at the existing way of formatting and have some concerns
- there is a lot of deserialization logic happening in the frontend and thus requiring a thick client, i.e. we will likely need to reimplement this logic in Java. The backend won't have to re-implement this code each time it tries to throw it, it can used a shared library, which is in c++ to format the message.
- In the future when we reformat the message we will have to change the payload on the wire / proto, i.e. its one of the reasons why we have this proto for propagating the error to the frontend. Wouldn't it be harder to maintain to have to keep adding new proto for each kind of error type?

```
message RayErrorInfo {
  oneof error {
    ActorDeathCause actor_died_error = 2;
    RuntimeEnvFailedContext runtime_env_setup_failed_error = 3;
    string error_message = 4;
  }
  // The type of error that caused the exception.
  optional ErrorType error_type = 11;
} 
```

Reply:
I think it's fine to have a thicker client for error deserialization (it's not a performance issue and the error formatting is pretty low effort). Even if we have a helper method to re-implement in the backend, I'm a bit concerned it will be more brittle because we can forget to call it. In this case, having a more structured proto actually seems worth it since it is well-defined what metadata the backend is required to provide, and we can unit-test that behavior more easily than checking error message content.

That said, what metadata do we need to pass back from the NodeManager? Some of the fields like worker ID and node ID are probably already available on the driver side, so we can populate them there. For other metadata that's only available on the NodeManager side, we can pass this back as a shorter formatted error string that gets embedded in the user-level exception message.

Reply:
How are we going to address the maintenance issue? Don't we need to re-implement and maintain this for each language like Python + Java?

To make this more robust we should add helper functions / factory method that generates the payload, i.e.

rpc::RayErrorInfo Errors::NewOutOfMemoryError(used_mem, total_mem, task_id, task_name, ......) {
     rpc::RayErrorInfo  error_info;
      std::stringstream worker_exit_message_ss;
          worker_exit_message_ss
              << "System memory low at node with IP " << latest_worker->IpAddress()
              << ". Used memory (" << used_bytes_gb << "GB) / total capacity ("
              << total_bytes_gb << "GB) (" << usage_fraction << ") exceeds threshold "
              << usage_threshold << ", killing latest task with name "
              << latest_worker->GetAssignedTask().GetTaskSpecification().GetName()
              << " and " << id_ss.str() << " to avoid running out of memory.\n"
              << "This may indicate a memory leak in a task or actor, or that too many "
                 "tasks are running in parallel.\n";
          std::string worker_exit_message = worker_exit_message_ss.str();
     error_info.error_message  = worker_exit_message_ss;
     return error_info
}

The message here is more flexible to change and more unit-testable. If we put it on the wire it is very hard to unit test since we don't have ways to test the raylet api directly, most such tests are done via the python test or mocked raylet client.

If we do it the other way where we populate parts of the message in the backend and then front it would looks something like this:


rpc::RayErrorInfo Errors::NewOutOfMemoryError(used_mem, total_mem, task_id, task_name, ......) {
     rpc::RayErrorInfo  error_info;
      std::stringstream worker_exit_message_ss;
          worker_exit_message_ss
              << "System memory low at node with IP " << latest_worker->IpAddress()
              << ". Used memory (" << used_bytes_gb << "GB) / total capacity ("
              << total_bytes_gb << "GB) (" << usage_fraction << ") exceeds threshold "
                 "tasks are running in parallel.\n";
          std::string worker_exit_message = worker_exit_message_ss.str();
     error_info.error_message  = worker_exit_message_ss;
     return error_info
}

and then in the frontend:

def new_oom_error(node_id, worker_id, backend_message) :
     error = OOMError()
     error.message =usage_threshold << 
     error.message = f"node id {node_id} ..... {backend_message killing latest task with name "
             {get_task_name()} and {task_id} to avoid running out of memory. This may indicate a "
            "memory leak in a task or actor, or that too many "
     return error
              

Wouldn't it be harder to understand and maintain and test if we do it separately like this?

Reply:
I don't see a strong reason to go either way. From the user perspective, it makes little difference, and the maintenance concerns seem OK in each solution.

It seems like a stylistic decision, so I'd vote to be consistent with how this is approached with other similar types of error messages.

Reply:
@stephanie-wang  can you let me know what should be part of the error message generated by the raylet and which part of the error message should be constructed on the frontend? This helps me figure out what fields should be added to the proto for the message that needs to be introduced.

The current message we sent back for OOM is 

```

std::stringstream worker_exit_message_ss;
          worker_exit_message_ss
              << "System memory low at node with IP " << latest_worker->IpAddress()
              << ". Used memory (" << used_bytes_gb << "GB) / total capacity ("
              << total_bytes_gb << "GB) (" << usage_fraction << ") exceeds threshold "
              << usage_threshold << ", killing latest task with name "
              << latest_worker->GetAssignedTask().GetTaskSpecification().GetName()
              << " and " << id_ss.str() << " to avoid running out of memory.\n"
              << "This may indicate a memory leak in a task or actor, or that too many "
                 "tasks are running in parallel.\n";
          std::string worker_exit_message = worker_exit_message_ss.str();

          std::stringstream error_log_ss;
          error_log_ss
              << worker_exit_message
              << "To find the highest memory consumers, use `ray logs raylet.out -ip "
              << latest_worker->IpAddress() << "`.\n"
              << "Consider provisioning more memory on this node or reducing task "
                 "parallelism by requesting more CPUs per task. "
              << "To adjust the eviction threshold, set the environment variable "
                 "`RAY_memory_usage_threshold_fraction` when starting Ray. "
              << "To disable worker eviction, set the environment variable "
                 "`RAY_memory_monitor_interval_ms` to zero.";


```




The current message we sent back for NodeDied is 

```

RAY_LOG(DEBUG) << "Failed to fetch task result with status "
                     << get_task_failure_cause_reply_status.ToString()
                     << " node id: " << addr.raylet_id << " ip: " << addr.ip_address;

```

Reply:
I think the only fields that aren't available on the driver already are the ones that have to do with physical resources on the node, like the current used memory or how long the task ran before it was killed. But I'll leave it up to you to decide which fields to send back, and we can always improve on this later.

The high-level bit as @ericl said is just to make sure that the error message is useful to the user. Right now, I feel like the OOM message doesn't really make sense as a task return error since the content was originally about logging the OOM kill event on the raylet, not explaining what happened for a specific task. The NodeDied error could also be more informative (explaining why it might've happened, what to do next, etc).

You could look at the various [ObjectLostErrors](https://github.com/ray-project/ray/blob/master/python/ray/exceptions.py#L371) for some examples of what sort of metadata and content to include.

Reply:
I see we have various ways of generating the message and not a clear consistent way of doing things. For example, ActorDiedErrorContext has all the fields populated from the backend, ObjectStoreFullError / OutOfDiskError doesn't have fields propagated, just things hard-coded. ObjectLostError doesn't have fields propagated from the backend either - the object ref is already available in the frontend. So I am still a little confused, and would prefer not to leave things up to me. In the interest of time it is best if you tell me what you want to avoid more back and forth. And it would be appreciated if we don't spend cycles getting everything perfect now so we could move on to the other (equally) important things like making OOM handling work better. 

In the latest comment what I am hearing is that we should have a different error message, which was not the comment at the beginning of this thread. So it is best if you help me understand what you want done from the start to avoid the back and forth. Thanks!

Reply:
@ericl   this PR is following 

> I don't see a strong reason to go either way. From the user perspective, it makes little difference, and the maintenance concerns seem OK in each solution.
> 
> It seems like a stylistic decision, so I'd vote to be consistent with how this is approached with other similar types of error messages.

Here is what we are doing today, we pass the message from the raylet through the task submitter

       ```
         } else if (reply.failure_type() ==
                           rpc::RequestWorkerLeaseReply::
                               SCHEDULING_CANCELLED_UNSCHEDULABLE) {
                  error_type = rpc::ErrorType::TASK_UNSCHEDULABLE_ERROR;
                  *(error_info.mutable_error_message()) =
                      reply.scheduling_failure_message();

```

Can you help me understand which part of this PR is not consistent ?

Reply:
Looking through the current exceptions, it seems like most of them use the pattern of passing some error context to Python and doing most of the formatting there, so I would prefer if we can make this the norm. But I didn't intend for this to block the PR, so if it's easier to format in C++, we can stick with that for now and follow up later to make all the errors consistent.

However, can you address the comment about improving the user-facing error messages for OutOfMemoryError and NodeDiedError before we merge? Currently they sound more like system log messages instead of task/worker-specific errors.